Magento2/Sniffs/PHP/AutogeneratedClassNotInConstructorSniff.php

```diff
@@ -0,0 +1,200 @@
+    {
+        return [T_FUNCTION, T_DOUBLE_COLON, T_USE];
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function process(File $phpcsFile, $stackPtr)
+    {
+        if ($phpcsFile->getTokens()[$stackPtr]['type'] === 'T_USE') {
+            $this->registerUses($phpcsFile, $stackPtr);
+        }
+        if ($phpcsFile->getTokens()[$stackPtr]['type'] === 'T_FUNCTION') {
+            $this->registerConstructorParameters($phpcsFile, $stackPtr);
+        }
+        if ($phpcsFile->getTokens()[$stackPtr]['type'] === 'T_DOUBLE_COLON') {
+            if (!$this->isObjectManagerGetInstance($phpcsFile, $stackPtr)) {
+                return;
+            }
+
+            $statementStart = $phpcsFile->findStartOfStatement($stackPtr);
+            $statementEnd = $phpcsFile->findEndOfStatement($stackPtr);
+            $equalsPtr = $phpcsFile->findNext(T_EQUAL, $statementStart, $statementEnd);
+
+            if (!$equalsPtr) {
+                return;
+            }
+
+            $variableInParameters = false;
+            if ($variable = $phpcsFile->findNext(T_VARIABLE, $equalsPtr, $statementEnd)) {
+                $variableName = $phpcsFile->getTokens()[$variable]['content'];
+                if ($variableName === '$this') {
+                    $variableName = $this->getNext($phpcsFile, $variable, $statementEnd, T_STRING)['content'];
+                }
+                foreach ($this->constructorParameters as $parameter) {
+                    $parameterName = $parameter['name'];
+                    if ($this->variableName($parameterName) === $this->variableName($variableName)) {
+                        $variableInParameters = true;
+                    }
+                }
+            }
+
+            if (!$variableInParameters) {
+                $next = $stackPtr;
+                while ($next = $phpcsFile->findNext(T_DOUBLE_COLON, $next + 1, $statementEnd)) {
+                    if ($this->getNext($phpcsFile, $next, $statementEnd, T_STRING)['content'] === 'class') {
+                        $className = $this->getPrevious($phpcsFile, $next, T_STRING)['content'];
+                    }
+                }
+
+                $className = $this->getClassNamespace($className);
+
+                $phpcsFile->addWarning(
+                    sprintf("Class %s needs to be requested in constructor, " .
+                        "otherwise compiler will not be able to find and generate these classes", $className),
+                    $stackPtr,
+                    self::ERROR_CODE
+                );
+            }
+        }
+    }
+
+    /**
+     * Check if it is a ObjectManager::getInstance
+     *
+     * @param File $phpcsFile
+     * @param int $stackPtr
+     * @return bool
+     */
+    private function isObjectManagerGetInstance(File $phpcsFile, int $stackPtr): bool
+    {
+        $prev = $phpcsFile->findPrevious(T_STRING, $stackPtr - 1);
+        $next = $phpcsFile->findNext(T_STRING, $stackPtr + 1);
+        if ($prev &&
```

Comment:
This can be simplified to
```php
        return $prev &&
               $next &&
               $phpcsFile->getTokens()[$prev]['content'] === 'ObjectManager' &&
               $phpcsFile->getTokens()[$next]['content'] === 'getInstance';
```